locations/automatic_spider_generator.py

```diff
@@ -0,0 +1,268 @@
+        data: str = None,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(url, headers, *args, **kwargs)
+        self.data = data
+
+    def __repr__(self):
+        parameters = []
+        if self.url:
+            parameters.append('url="{}"'.format(self.url))
+        if self.headers:
+            parameters.append("headers='{}'".format(self.headers))
+        if self.data:
+            parameters.append("data='{}'".format(self.data))
+        return "DetectionRequestRule({})".format(", ".join(parameters))
+
+    def __bool__(self):
+        if self.url or self.headers or self.data:
+            return True
+        else:
+            return False
+
+
+class DetectionResponseRule(DetectionRule):
+    js_objects: dict = {}
+    xpaths: dict = {}
+
+    def __init__(
+        self,
+        url: str = None,
+        headers: str = None,
+        js_objects: dict = {},
+        xpaths: dict = {},
+        *args,
+        **kwargs,
+    ):
+        super().__init__(url, headers, *args, **kwargs)
+        self.js_objects = js_objects
+        self.xpaths = xpaths
+
+    def __repr__(self):
+        parameters = []
+        if self.url:
+            parameters.append('url="{}"'.format(self.url))
+        if self.headers:
+            parameters.append("headers='{}'".format(self.headers))
+        if self.js_objects:
+            parameters.append("js_objects={}".format(self.js_objects.__repr__()))
+        if self.xpaths:
+            parameters.append("xpaths={}".format(self.xpaths.__repr__()))
+        return "DetectionResponseRule({})".format(", ".join(parameters))
+
+    def __bool__(self):
+        if self.url or self.headers or self.js_objects or self.xpaths:
+            return True
+        else:
+            return False
+
+
+class AutomaticSpiderGenerator:
+    detection_rules: list[DetectionRule] = []
+
+    @staticmethod
+    def generate_spider_code(spider: Spider) -> str:
+        """
+        Generate source code representation of a spider class, where
+        the generated source code is intended to be executable
+        without further changes being required.
+        :param spider: Spider class which should have a source code
+                       representation generated for.
+        :return: Generated source code (multi-line) of spider.
+        """
+        imports_list = ""
+        superclasses = []
+        for spider_base in spider.__bases__:
+            imports_list = "{}from {} import {}\n".format(imports_list, spider_base.__module__, spider_base.__name__)
+            superclasses.append(spider_base.__name__)
+        superclasses_list = ", ".join(superclasses)
+
+        spider_code = "{}\n\nclass {}({}):{}".format(
+            imports_list,
+            spider.__name__,
+            superclasses_list,
+            AutomaticSpiderGenerator.generate_spider_attributes_code(spider),
+        )
+        return spider_code
+
+    @staticmethod
+    def generate_spider_attributes_code(
```

Comment:
After looking at what scrapy does; longer term is it better to refactor this more towards "template and variables" as per https://github.com/scrapy/scrapy/blob/master/scrapy/templates/spiders/crawl.tmpl ?

I would say because this is working and useful right now that it should be merged as is and this is absolutely a "what do we look at next" problem.

But... https://github.com/scrapy/scrapy/blob/master/scrapy/commands/genspider.py#L147 has some nice bits to it; though I am not sure if 
https://github.com/scrapy/scrapy/blob/master/scrapy/utils/template.py#L18C22-L18C30
supports conditional logic/etc; which this code certainly does.

Reply:
I've had another look at Scrapy templates and am OK to proceed with switching to use of templates. It's really just https://docs.python.org/3/library/string.html#template-strings with a few reserved substitution placeholders defined in `_generate_template_variables` (https://github.com/scrapy/scrapy/blob/master/scrapy/commands/genspider.py) that we should avoid reusing.

I think what we need to figure out whether the following properties should exist:
1. Spiders generated from templates should be immediately executable with only the substitution placeholders provided from command line or from variables extracted from automatic spider generation. No need to manually comment/uncomment lines in the generated spider.
2. If some substitution placeholders are optional (for example a "api_filter" attribute of a spider class) then the generated spiders should contain those optional attributes commented out, with a default state of `None` (for example). Before committing the generated spider, it is up to the user to remove these unnecessary commented out lines, or uncomment the lines and set a real value if this is desired.
3. Generated spiders should include default pass-through `pre_process_data, `post_process_item`, `parse_opening_hours`, etc functions which are overridden from parent classes. This way it is easy for the user to go ahead and add custom logic to these overridden functions, or delete unused functions that don't need overriding.

Reply:
Moved to New feature request ticket #10198